[Refactor] Apply More BitField

Minty-Meeo opened the pull request:
Replace as many bitfield structs with Common's BitField as was simple to do.

Minty-Meeo commented:
This PR is ready for more eyes to look at it.  I still want to go around and polish it up a bit, so it is a draft for now.
This refactor bring a ton of unions across the codebase up to the modern day.  I had a ton of trouble getting bitfields in ButtonData classes to cooperate, though.  The liberal use of BitCastPtr clashes with the restrictive properties of the BitField class.  I am considering rolling back all controller changes rather than leaving in a half-finished conversion.

Minty-Meeo commented:
I also want to ask, what is the best way to look at the assembly of the compiled program using VS Code on Linux?  I can't seem to find a disassembly window like Visual Studio has.  I want to be absolutely sure that the compiled code is in fact the same as using simple bitfield structs.

Minty-Meeo commented:
oh right I forgot to expand the stdio output

Pokechu22 commented:
The fifoci diffs are (presumably) due to this PR being based on an older revision, while the version of Mesa on the radeon builder was updated from 21.3.7 for 5.0-16326 ([build log](https://dolphin.ci/#/builders/8/builds/1761)) to 22.0.1 for 5.0-16331 ([build log](https://dolphin.ci/#/builders/8/builds/1762)).  They should go away if you rebase on master (though they've also disappeared for now since it's comparing against the previous push to this PR).

I'm not sure what the best way to look at assembly is, but `objdump` should work at least.

Minty-Meeo commented:
Common's BitField lacking copy-assignment is such a drag.  I don't suppose restricting using it in unintended ways via code review is on the table?

Minty-Meeo commented:
I think I have found an acceptable compromise with new DOLPHIN_DEPRECATED and DOLPHIN_DEPRECATED_MSG decorators.  Improper usage of copy-assignment can now be detected by searching for deprecated function warnings containing the message "If you can see this warning, you might be using BitField the wrong way."  It does not seem to trigger from implicit copy-assignment, just explicit usage (the exact kind we want to avoid).  I'll be doing more testing as I go.

shuffle2 commented:
windows build is failing because you have an extra leading `_`, but it will still fail because the deprecated warning is ...a warning, and all warnings are errors on windows build.

I think you should decide to just use the copy assignment or not.

Minty-Meeo commented:
Copy assignment is necessary for copying-assigning structs which contain BitField unions, or else there are un-copyable sections in the struct.  Copying BitField to BitField gives unexpected behavior.  User-defined copy-assignment would break tons of old code and would also probably be very slow.  BitField is kind of asking for the impossible.  Without allowing default copy-assignment, it cannot fully replace bit structs.

Minty-Meeo commented:
[https://github.com/dolphin-emu/dolphin/blob/260d3faba8ba5b706a443ed6e2b6b39b972f9296/Source/Core/Core/PowerPC/JitArm64/JitArm64_LoadStore.cpp#L598]
Just noticed this while reviewing my changes so far.  0xfc00'ffff?  Is this a typo?  I'm surprised it compiles fine.

sepalani commented:
@Minty-Meeo 
If you're referring to the single quote, I think that's intended as `'` can be used as a digit separator:
https://en.cppreference.com/w/cpp/language/integer_literal

Minty-Meeo commented:
Bit struct changed to Common's BitFields in union:
 - ActionReplay::ARAddr
 - AudioInterface::AICR
 - DSP::UARAMCount
 - DSP::UAudioDMAControl
 - DSP::ARAM_Info
 - DSP::UDSPControl
 - ExpansionInterface::CEXIChannel::UEXI_STATUS
 - ExpansionInterface::CEXIChannel::UEXI_CONTROL
 - ExpansionInterface::CEXIMic::UStatus
 - VideoInterface::UVIVerticalTimingRegister
 - VideoInterface::UVIDisplayControlRegister
 - VideoInterface::UVIHorizontalTiming0
 - VideoInterface::UVIHorizontalTiming1
 - VideoInterface::UVIVBlankTimingRegister
 - VideoInterface::UVIBurstBlankingRegister
 - VideoInterface::UVIFBInfoRegister
 - VideoInterface::UVIInterruptRegister
 - VideoInterface::UVILatchRegister
 - VideoInterface::PictureConfigurationRegister
 - VideoInterface::UVIHorizontalScaling
 - VideoInterface::UVIFilterCoefTable3
 - VideoInterface::UVIBorderBlankRegister
 - VideoInterface::UVIDTVStatus
 - VideoInterface::UVIHorizontalStepping
 - WiimoteCommon::OutputReportGeneric
 - WiimoteCommon::OutputReportRumble
 - WiimoteCommon::OutputReportEnableFeature
 - WiimoteCommon::OutputReportLeds
 - WiimoteCommon::OutputReportMode
 - WiimoteCommon::OutputReportRequestStatus
 - WiimoteCommon::OutputReportWriteData
 - WiimoteCommon::OutputReportReadData
 - WiimoteCommon::OutputReportSpeakerData
 - WiimoteCommon::ButtonData
 - WiimoteCommon::InputReportStatus
 - WiimoteCommon::InputReportReadDataReply
 - WiimoteCommon::AccelCalibrationPoint
 - WiimoteCommon::AccelCalibrationData
 - WiimoteEmu::Classic
 - WiimoteEmu::Classic::ButtonFormat
 - WiimoteEmu::Drums::DataFormat
 - WiimoteEmu::Guitar::DataFormat
 - WiimoteEmu::Nunchuk::ButtonFormat
 - WiimoteEmu::Turntable::DataFormat
 - WiimoteEmu::UDrawTablet::DataFormat
 - WiimoteEmu::MotionPlus::DataFormat
 - UGeckoInstruction
 - CommandProcessor::UCPStatusReg
 - CommandProcessor::UCPCtrlReg
 - CommandProcessor::UCPClearReg

No BitFields necessary:
 - AudioInterface::AIVR
 - SerialInterface::UCommand
 - VideoInterface::UVIFilterCoefTable4

Other changes:
 - Header "Common/Deprecated.h" created.  It contains DOLPHIN_DEPRECATED and DOLPHIN_DEPRECATED_MSG macros.
 - BitField changes:
    - Default copy-assignment restored.  It was formerly deleted to forbid using it, as it produces an unexpected result when operator= is used.  However, this also deletes copy-assignment for all classes containing BitField unions, which is undesireable for many reasons listed in an updated comment.  Using operator= in a way which produces an unexpected result will now result in a deprecated function warning.  Copy-assignment itself does not produce this warning.  Using operator= with classes containing BitField unions does not appear to produce a deprecated function warning.
    - Added advice for working around BitField to BitField operator=.
    - SetValue method added
    - More assignment operator overloads added
 - BitFieldArray has had copy-assignment changed ion the same way as BitField.
 - Enumeration SerialInterface::EdirectCommands::CMD_NULL added to simplify code a bit.
 - Restored documentation messed up by clang-format in ExpansionInterface::CEXIChannel::UEXI_STATUS
 - UGeckoRegister has been given a great glow-up
    - Redundant bit fields removed
    - Some documentation added
    - BitFields are organized into categories
    - Pretty proud of this change
 - Removed bogus code for a non-existent gprC

Minty-Meeo commented:
Found even more but now I'm touching ShaderGen code and I really hope I don't screw something up.

Minty-Meeo commented:
I don't want to speak too soon, but I think I have finally located and changed every bit struct I want to tackle with this PR.  Wow, this has gotten huge.

Minty-Meeo commented:
Wait a minute, all this time I thought RFC was short for "Refactor Code".  Guess I'll edit the PR name.

Pokechu22 commented on Source/Core/Common/BitField.h:
```diff
@@ -133,21 +140,69 @@ struct BitField
+  // solution is to allow copy-assignment to exist, albeit with a warning so that code-review is not
+  // impossible.
+  DOLPHIN_DEPRECATED_MSG("If you can see this warning, you might be using BitField the wrong way.")
+  constexpr BitField& operator=(const BitField&) = default;
```
Have you made sure that the deprecated message actually shows up?  When I was working on a similar thing that eventually ended up as dolphin-emu/hwtests#47, the warning didn't show up on GCC when `default` was used, but maybe I was doing something wrong?

Minty-Meeo replied:
I have caught many unexpected behavior copy-assignments using the error message since I made that change to progress further.

Minty-Meeo replied:
One interesting thing I've seen is that copy-assignment between BitFields with different template args does not seem to be an issue.  Would probably be best to still use .Value() just to be consistent.

Pokechu22 replied:
Hmm, I re-tested it, and both it and the standardized [`[[deprecated("Message")]]`](https://en.cppreference.com/w/cpp/language/attributes/deprecated) do generate warnings properly on GCC and Clang (even the version of GCC used by devkitpro).  But no warning is generated on MSVC, and Clang also generates a warning for the implicit usage.  See https://godbolt.org/z/6EWEMfvGe and the [pr-osx-universal buildlog](https://dolphin.ci/#/builders/26/builds/2935/steps/4/logs/stdio) near line 1242.

CC @neobrain

Minty-Meeo replied:
I did not realize [[deprecated]] was a language standard.  I think I will get rid of the header I created and just use that.

Minty-Meeo replied:
Clang Format is absolutely wrecking the [[deprecated("message")]] one.
[https://stackoverflow.com/questions/45740466/clang-format-setting-to-control-c-attributes](url)

neobrain replied:
What exactly is the idea here? Is there going to be a bunch of "deprecated: ..." spam by default that can be ignored for "blessed" uses of this operator (i.e. those needed to copy unions), and someone needs to keep a close eye on making sure no new warnings slip through code review in the future?

neobrain replied:
What exactly is the idea here? Is there going to be a bunch of "deprecated: ..." spam by default that can be ignored for "blessed" uses of this operator (i.e. those needed to copy unions), and someone needs to keep a close eye on making sure no new warnings slip through code review in the future?

Or is the hope to find a variant of this that only triggers the warning for explicit BitField-to-BitField copies but not with implicit copying through unions?

iwubcode replied:
FWIW, I'm not a huge fan of the deprecated tag being used in this way.  There's too much book keeping.  Feels like a design smell.

This is the only way I could see to do what you want, but it's very ugly in a _different_ way:

```
#include <algorithm>

template <typename Parent>
class Foo
{
public:
    Foo() = default;
    void Copy(const Foo<Parent>& other)
    {
        // Do relevant copying...
    }
    int i;
private:
    friend Parent;

    // rule of 5
    Foo& operator=(const Foo& other) = default;
    Foo& operator=(Foo&& other) = default;
    Foo(const Foo& other) = default;
    Foo(Foo&& other) = default;
};

union Bar
{
    Foo<Bar> m_foo;
    Foo<Bar> m_foo2;
};

int main()
{
    Bar bar1;
    Bar bar2;
    bar2.m_foo = bar1.m_foo;  // compiler error
    bar1 = bar2;  // works
    bar1.m_foo.Copy(bar2.m_foo);  // works
    std::swap(bar1, bar2);  // works
    return 0;
}
````

Minty-Meeo commented:
Is this another case of needing to rebase, or did I finally hit a snag tampering with ubershader code?

Pokechu22 commented on Source/Core/VideoCommon/PixelShaderGen.h:
```diff
@@ -167,7 +167,8 @@ struct pixel_shader_uid_data
   LightingUidData lighting;
-  BitFieldArray<0, 8, 1, bool, u8> texMtxInfo_n_projection;  // 8x1 bit
+  // Stored separately to guarantee that the texMtxInfo struct is 8 bits wide
+  BitFieldArray<0, 8, 1, u8> texMtxInfo_n_projection;
```
I'm pretty sure this should be `BitFieldArray<0, 1, 8, u8>`, as you want 8 fields where each field is one bit, not 1 field that's 8 bits.  This might be the source of your fifoci diffs.  (Though I'd also recommend doing a full rebase just to be certain).  This also applies to VertexShaderGen.h.

Minty-Meeo replied:
That would explain the static assert I was getting for the 16 bit sized one

Pokechu22 commented:
You should also be able to get rid of [the explicit copy constructors in RenderState.h](https://github.com/dolphin-emu/dolphin/blob/902e45b6b76d3c7aeab478a954212f841530ed80/Source/Core/VideoCommon/RenderState.h#L25-L37), hopefully.

merryhime commented on Source/Core/Common/Deprecated.h:
```diff
@@ -0,0 +1,12 @@
+
+#pragma once
+
+#ifdef _WIN32
```
As this is compiler specific rather than OS specific, `_MSC_VER` would likely be more appropriate.

Minty-Meeo replied:
Looks line Inline.h could use that, too.

Minty-Meeo replied:
Should the \_\_attribute\_\_ decorator be only for GNU, or is it safe for it to always fall through to that option?

Pokechu22 replied:
Clang also supports [`always_inline`](https://clang.llvm.org/docs/AttributeReference.html#always-inline-force-inline) and [`deprecated`](https://clang.llvm.org/docs/AttributeReference.html#always-inline-force-inline).

Pokechu22 replied:
(however, I think clang does define `__GNUC__` so checking that way is fine)

neobrain commented:
The design conflicts observed in this PR are one of the reasons why for ongoing projects I've moved away from the `union`-based approach. My current design represents bit fields as member functions instead, which trivially preserves default constructors and operators of the parent type.

Decaf has adopted a variant of my design a while ago ([implementation](https://github.com/decaf-emu/decaf-emu/blob/master/src/common/bitfield.h) & [example](https://github.com/decaf-emu/decaf-emu/blob/master/src/libgpu/latte/latte_registers_pa.h)), but it uses macros, which come with their own set of gotchas.

My [current design](https://gist.github.com/neobrain/8446bc2dbbf7dee0bc126f20b1b35fcd#file-bit_field_test-cpp) works without macros, but other than the linked example code I haven't gotten around to publishing its implementation yet. Should be relatively straightforward to figure out how it's done by comparing against decaf though!

Minty-Meeo commented:
I appreciate all of the feedback and help.  I've not abandoned this PR yet, I've just been working hard to make something that will hopefully satisfy everybody.  While it's not completely finished yet, I'd like to share what I have made so far:
https://godbolt.org/z/hPsc77xeP
BitFieldArrays have been giving me a lot of heartache.  I'm not so familiar with some of the concepts I'm having to implement for them, but I'm trying my best!

Minty-Meeo commented:
This is certainly a "Works on my machine" moment, but finally my work is ready to show.  It definitely needs a second and a third pass to polish things up.  It appears nested templated classes are either something I don't understand or something compilers really don't like to do, hence why the fmt::formatter specialization tripped up every platform except x86-ubu

Pokechu22 commented:
I haven't tested it, but try this instead:

```C++
template <typename StorageType, typename FieldType, std::size_t start, std::size_t width>
struct fmt::formatter<typename BitField2<StorageType>::FixedView<FieldType, start, width>>
```

Minty-Meeo commented:
See, that's what I thought too, but all I get is this:
[{
	"resource": ".../dolphin/Source/Core/Common/BitField2.h",
	"owner": "cmake-build-diags",
	"severity": 8,
	"message": "template parameters not deducible in partial specialization:",
	"source": "GCC",
	"startLineNumber": 467,
	"startColumn": 13,
	"endLineNumber": 467,
	"endColumn": 1000,
}]
I looked it up and other errors I got while messing with it, and started seeing topics about compiler bugs and it was quite worrying.  I think for now I will un-nest the classes like the old BitField class did, peppering in lots of friend statements.

neobrain commented on Source/Core/Common/BitField2.h:
```diff
@@ -0,0 +1,544 @@
+#pragma once
```
Previous BitField.h was BSD licensed. Would you like to license this as BSD as well so other projects can integrate this more easily, or do you prefer keeping it as GPL?

Minty-Meeo replied:
I actually have been wondering about this lately.  If I were to contribute wholly original code to Dolphin Emulator, what rights do I retain?  Am I able to license it under a more permissive license than GPL in one of my own repositories, but still give Dolphin Emulator full permission to use it?  I was beginning to cook up a better MetroWerks Linker Map reader before this PR spiraled into the crazy thing it has now become.  Software licensing is something I'm still really unfamiliar with, so I'd appreciate any guidance.

Minty-Meeo replied:
In some projects in the past where I pretended to understand the licensing, I would use my online handle instead of my real name for the copyright info.  I'm not going to pretend I have perfect anonymity online, but I prefer to use my pseudonym.  Do you have any advice for this as well?

Minty-Meeo replied:
MIT and Unlicense are the only licenses I really am familiar with.  I'm happy with whatever, really, so long as it doesn't somehow get locked up in legal limbo.

neobrain replied:
(The usual disclaimer applies for this whole comment: I'm not a lawyer, and this is not legal advice)

>  If I were to contribute wholly original code to Dolphin Emulator, what rights do I retain

You retain all rights to the code you wrote yourself. Licensing just gives others permission to do things with your code, such as copying and modifying it.

>  Am I able to license it under a more permissive license than GPL in one of my own repositories, but still give Dolphin Emulator full permission to use it?

Yes, you can publish the same code (assuming you wrote all of it yourself) at a different place under a different license. The license you provide via the Dolphin source code is not affected by this.

BitField.h's license header is a bit funky in this regard. It's licensed as BSD (as mentioned starting from line 4), but then relicensed as GPL since that's what Dolphin as a whole uses. In hindsight I'd probably omit the explicit GPL notice, but the idea is that if people want to use BitField.h in other projects, they can do so without worrying about Dolphin being GPL-licensed as a whole.

> In some projects in the past where I pretended to understand the licensing, I would use my online handle instead of my real name for the copyright info. I'm not going to pretend I have perfect anonymity online, but I prefer to use my pseudonym. Do you have any advice for this as well?

People more active in the project than I should comment on this, though I'll note that using a pseudonym makes it difficult for project owners to verify your contribution doesn't legally poison the repository (such as, for example, submitting code that actually belongs to your employer).

Pokechu22 replied:
In my opinion (also not a lawyer), it'd be fine to use your pseudonym or even just "Dolphin emulator project".

`BitSet.h` also has a different license, but for it all that's listed is `SPDX-License-Identifier: CC0-1.0` ([Creative Commons Public Domain Declaration](https://creativecommons.org/share-your-work/public-domain/cc0)), which functions similarly to the Unlicense (e.g. both do not have an attribution requirement, so no name is listed).  For that file that's the only license listed, which is probably preferable.

neobrain commented on Source/Core/Common/BitField2.h:
```diff
@@ -0,0 +1,544 @@
+    {
+      const std::size_t rshift = 8 * sizeof(StorageType) - width;
+      const std::size_t lshift = rshift - start;
+      // Necessary to choose between arithmatic and logical right-shift.
```
```suggestion
      // Necessary to choose between arithmetic and logical right-shift.
```

shuffle2 commented:
the windows build is broken because 7b265ee3d11e2adfddd8fe2b3dc0a3c3a28ee3e7 reverts the Qt submodule update, please remove that.

shuffle2 commented:
now it fails because you need to rebase on master. can you just rebase and drop the stuff touching Qt submodule at all?

Minty-Meeo commented:
Idk how to erase a commit buried beneath other commits, but I do know how to squash commits to tidy this all up.

Pokechu22 commented:
Use `git rebase -i master` and then delete the lines for the commits you don't want.

Minty-Meeo commented:
Super excited to share this update.  Dolphin Emulator's needs have really put my new implementation through it's paces, and two days ago I hit a real snag while updating VideoInterface.h.  Several URegs in this header actually use unions in a more traditional way, and for a while I could not figure out how to support this until I realized what was missing.  I had built BitField2 to work entirely for polymorphism via inheritance, but this situation required polymorphism via composition!  Now, on top of the FIELD and FIELDARRAY macros, FIELD_IN and FIELDARRAY_IN macros are provided which allow you to specify a member which is a BitField2 to be the host of that View.
This big change should also hopefully make re-doing the ShaderGen bitfields possible.  I knew they were going to complicate things, and I hope this is all I need to make them work.

Minty-Meeo commented:
I don't know what is breaking ARM64's PowerPCTest unit test and I don't know how to find out.  Buildbot's log suggests "Use '--rerun-failed --output-on-failure' to re-run the failed cases verbosely." for the pr-osx-universal one.  I think something like that sounds like a good idea.

Minty-Meeo commented:
Is such a thing possible in C++?
```
    constexpr operator FieldType() const { return Get(); }
    if constexpr (std::is_enum<FieldType>())
    {
      using UnderlyingFieldType = typename std::underlying_type<FieldType>::type;
      constexpr operator UnderlyingFieldType() const { return Get(); }
    }
````

Minty-Meeo commented:
Progress came to a screeching halt a few days ago when I came upon the helper function, BitfieldExtract, in ShaderGenCommon.h.  Once again I ran into the issues of deducing types of a nested template class.  After some experimentation and thinking, I decided that un-nesting my implementation was the way forward.  I'll sort of miss how clean it was in certain ways, but what I've turned it into is respectable in its own ways.
This new approach is now 100% composition rather than inheritance, so all of my work porting existing code to the new implementation will need a re-do.  As well, the wrapper class for the underlying storage, previously called BitField2, is gone.  The various field view classes now work on raw integral values.  I think that this is better overall, though.  The existing URegs will be more similar to how they originally were, with weird things like default member initializer remaining intact.

Minty-Meeo commented:
It probably needed to happen either way.  You'll see in BitField3.h I finally got the fmt::formatter specializations working with ease!

Rumi-Larry commented on Source/Core/Common/BitFieldView.h:
```diff
@@ -0,0 +1,581 @@
+#pragma once
```
The legal header is missing.

Minty-Meeo replied:
Saving that for later.  Still haven't decided if I'll apply BSD or something else.

Minty-Meeo commented:
Uh oh, that's bad FIFO changes.

Minty-Meeo commented:
lmao I really messed it up this time

Minty-Meeo commented:
Realized while fixing my mistake that uid_data.lighting_data is basically a two-dimensional bitfield array.  Going to resist the urge to implement such a thing in BitFieldView, even though it'd totally be possible lol.

Minty-Meeo commented:
I have combed through every change I made since rendering stopped working, and I cannot find the needle in the haystack.  I think my only solution is to roll back to when I finished porting the bpmem structs, fix the two compiling mistakes left in that commit, and re-do it all.  I'll keep a backup of the progress up to this point, of course.

Minty-Meeo commented:
Software Rendering works fine.  OGL works with specialized, hybrid ubershaders, skip drawing, but for some reason exclusive ubershaders leaves everything totally black.  Fog appears, but that's about it.  Vulkan is the same story.  Anything I might be doing on accident on my end to cause this?  A clean build works fine with exclusive ubershaders.

Rumi-Larry commented:
I think you should split these changes into parts, so it is obvious what may cause a regression and is easier to review. It would also be a great opportunity to start with concise commits rather than piling them up for every minor change.
That's of course not to downplay the work you have already done; to the contrary, since you already done most of the work it would be trivial to do it in smaller chuncks.

Minty-Meeo commented:
I found the needle.  Very mad, because of course the clues were all there.
I changed some u32s in PixelShaderConstants to be bools.  The comments indicate these are for ubershaders only, which should have been my second clue.  My first clue was that these were u32s despite clearly meaning to be bools, and there are even comments indicating they should be treated as bools.  I don't even know why these being bools makes ubershaders bust.  Might have something to do with bool being a 32-bit sized type, or at least that's what it appears from things like VKBool32 that I saw around.

Minty-Meeo commented:
What is up with the spirv_cross external for Android?  Is Android broken at the moment?

Minty-Meeo commented:
And please, for the pr-osx-universal build-bot, add the "--rerun-failed --output-on-failure" flags to re-run the failed unit tests verbosely.

shuffle2 commented:
> What is up with the spirv_cross external for Android? Is Android broken at the moment?

rebase on master

Minty-Meeo commented:
One thing that has been on my mind during this PR is the potential for supporting floating point types as host types.  I've whipped up a quick hack of the BitFieldView header in Compiler Explorer that does just that.  The question is, does anyone even want this?
https://godbolt.org/z/avPsKTrvr

OatmealDome commented:
@Minty-Meeo

Here's the unit test failure on pr-osx-universal: https://pastebin.com/Y25H6PQd

Since the build script just runs `BuildMacOSUniversalBinary.py`, I think the relevant flags can just be placed in that. I'll look into it.

Minty-Meeo commented:
I made the 5 bit wide "FPRF" field in UReg_FPSCR a bool :skull: now I know why DuckStation added a static assert for bools in particular.

Minty-Meeo commented:
Is it fine for everything in BitFieldView.h to exist in the global namespace, or should it be put into the Common namespace?

iwubcode commented:
I know this is still being worked on and I'm hardly one to talk with my giant PRs but I wonder if this PR could be split out so that individual sub-systems were "upgraded" to the new bitfield a little at a time?  Much like how lioncash did for converting over to the fmt logger.

Minty-Meeo commented:
If it is really necessary, I can probably split this PR into chunks once I've given everything one last quality assurance pass.

shuffle2 commented:
@Minty-Meeo what is your development environment? most/all have built-in way to apply the formatting we specify. We also have [a script in Tools to do it](https://github.com/dolphin-emu/dolphin/blob/master/Tools/lint.sh)

Minty-Meeo commented:
I use Visual Studio Code w/ auto format on save.  However, it does *not* auto-format when using the symbol renaming feature.  I usually just kinda cross my fingers that nothing is left un-formatted.

shuffle2 commented:
The lint script does work under wsl, however it can be a bit slow if the sources are on the host fs. I made a PR to fix that: #10733 

so, you should be able to use something like 
`/mnt/c/src/dolphin$ ./Tools/lint.sh --force HEAD`

..actually, idk if you're using vscode on windows, but i assumed that's maybe why you weren't using the script.

Minty-Meeo commented:
I'm on Manjaro Linux.  Also I think the FifoCl is having a stroke.

Minty-Meeo commented:
I have learned about the very cool and neat EnumMap class.  It is handy for removing silliness like u32(ArbitraryField().Get()) that arises when using a std::array with enumerated values.  Unfortunately, I sort of ran into a wall when trying to update the following std::array to a Common::EnumMap:
https://github.com/dolphin-emu/dolphin/blob/3cd82b619388d0877436390093a6edc2319a6904/Source/Core/VideoCommon/RenderState.cpp#L165-L202

Trying it like so gives me the error `no instance of constructor "Common::EnumMap<V, last_member, <unnamed>>::EnumMap [with V=LogicOpApproximation, last_member=LogicOp::Set, <unnamed>=LogicOp]" matches the argument list`
```
  static constexpr Common::EnumMap<LogicOpApproximation, LogicOp::Set> approximations = {{
      {false, SrcBlendFactor::Zero, DstBlendFactor::Zero},            // CLEAR
      {false, SrcBlendFactor::DstClr, DstBlendFactor::Zero},          // AND
      {true, SrcBlendFactor::One, DstBlendFactor::InvSrcClr},         // AND_REVERSE
      {false, SrcBlendFactor::One, DstBlendFactor::Zero},             // COPY
      {true, SrcBlendFactor::DstClr, DstBlendFactor::One},            // AND_INVERTED
      {false, SrcBlendFactor::Zero, DstBlendFactor::One},             // NOOP
      {false, SrcBlendFactor::InvDstClr, DstBlendFactor::InvSrcClr},  // XOR
      {false, SrcBlendFactor::InvDstClr, DstBlendFactor::One},        // OR
      {false, SrcBlendFactor::InvDstClr, DstBlendFactor::InvSrcClr},  // NOR
      {false, SrcBlendFactor::InvDstClr, DstBlendFactor::Zero},       // EQUIV
      {false, SrcBlendFactor::InvDstClr, DstBlendFactor::InvSrcClr},  // INVERT
      {false, SrcBlendFactor::One, DstBlendFactor::InvDstAlpha},      // OR_REVERSE
      {false, SrcBlendFactor::InvDstClr, DstBlendFactor::InvSrcClr},  // COPY_INVERTED
      {false, SrcBlendFactor::InvDstClr, DstBlendFactor::One},        // OR_INVERTED
      {false, SrcBlendFactor::InvDstClr, DstBlendFactor::InvSrcClr},  // NAND
      {false, SrcBlendFactor::One, DstBlendFactor::One},              // SET
  }};
```
I get the same error when removing one layer of closing braces, with an additional error: `could not convert ‘{{false, SrcBlendFactor::Zero, DstBlendFactor::Zero}, {false, SrcBlendFactor::DstClr, DstBlendFactor::Zero}, {true, SrcBlendFactor::One, DstBlendFactor::InvSrcClr}, {false, SrcBlendFactor::One, DstBlendFactor::Zero}, {true, SrcBlendFactor::DstClr, DstBlendFactor::One}, {false, SrcBlendFactor::Zero, DstBlendFactor::One}, {false, SrcBlendFactor::InvDstClr, DstBlendFactor::InvSrcClr}, {false, SrcBlendFactor::InvDstClr, DstBlendFactor::One}, {false, SrcBlendFactor::InvDstClr, DstBlendFactor::InvSrcClr}, {false, SrcBlendFactor::InvDstClr, DstBlendFactor::Zero}, {false, SrcBlendFactor::InvDstClr, DstBlendFactor::InvSrcClr}, {false, SrcBlendFactor::One, DstBlendFactor::InvDstAlpha}, {false, SrcBlendFactor::InvDstClr, DstBlendFactor::InvSrcClr}, {false, SrcBlendFactor::InvDstClr, DstBlendFactor::One}, {false, SrcBlendFactor::InvDstClr, DstBlendFactor::InvSrcClr}, {false, SrcBlendFactor::One, DstBlendFactor::One}}’ from ‘<brace-enclosed initializer list>’ to ‘const Common::EnumMap<BlendingState::ApproximateLogicOpWithBlending()::LogicOpApproximation, LogicOp::Set>’`

The best answer I can find is that EnumMap would need to be an aggregate type in order to support this, but since it has a special constructor to enforce its length, it is not an aggregate type.  Is there anything that can be done about this?

Pokechu22 commented:
This is a case of C++ having strange rules regarding brackets... I think you can write `LogicOpApproximation{false, SrcBlendFactor::Zero, DstBlendFactor::Zero}`. You can also add a function to force it to deduce to the correct type, which I did in [some of the vertex loaders](https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/VideoCommon/VertexLoader_Position.cpp#L119-L125).

Minty-Meeo commented:
I feel as though my changes in Tev.h to eliminate ().Get() and casts is out of the scope of this PR.

Pokechu22 commented:
I think I've made similar changes in #10549.

Minty-Meeo commented:
Pretty much the same as I envisioned.  I'll leave that to you, then.

Minty-Meeo commented:
Okay, save for a minor change or two, this behemoth of a project is finally ready to enter its final stage: splitting into parts.  What would be the best way to do it?
1) Multiple new Pull Requests, single commit each.
2) Single new Pull Request, multiple commits.
3) Rewrite PR 10610's history, multiple commits.

Pokechu22 commented:
In my opinion: Multiple new pull requests, one that introduces the new changes to `BitField.h` and `BitFieldView.h` (assuming you can make those changes without needing to change other files), and then several that include a copy of that first PR and then one or more additional commits related to a part of Dolphin (e.g. SI, the graphics code, etc.), which one PR per region where it's feasible to do so (and if you can split these into smaller commits, such as changes to `CPMemory.h`, `BPMemory.h`, etc within the same pull request, that'd be good, but it's not something that's essential).

If you include a copy of the first pull request's commit in the other pull requests, then when that first pull request is merged `git rebase` should automatically skip it (and in the mean time, things will build correctly and each pull request can get reviews).

Minty-Meeo commented:
The initial commit will introduce BitFieldView, it's unit test, touch EnumMap to add BitFieldView support, and do CMakeList housekeeping.
Files that Gekko.h influences are pretty tightly wound together.  It's pretty unrealistic to split them.
The BPMemory/CPMemory conglamorate *might* be possible to split into two parts, but it would be difficult and I'd rather not.
Everything else is pretty independent.
BitField.h I assume will be removed in one final commit once nothing uses it, so I won't bother putting anything new into it.
